libs/labelbox/src/labelbox/data/serialization/ndjson/label.py

```diff
@@ -24,23 +23,68 @@
                        NDPromptText]
 
 
-class NDLabel(pydantic_compat.BaseModel):
-    annotations: List[AnnotationType]
+class NDLabel(BaseModel):
+    annotations: List[SubclassRegistryBase]
+    
+    def __init__(self, **kwargs):
+        # NOTE: Deserialization of subclasses in pydantic is difficult, see here https://blog.devgenius.io/deserialize-child-classes-with-pydantic-that-gonna-work-784230e1cf83
+        # Below implements the subclass registry as mentioned in the article. The python dicts we pass in can be missing certain fields
+        # we essentially have to infer the type against all sub classes that have the SubclasssRegistryBase inheritance. 
+        # It works by checking if the keys of our annotations we are missing in matches any required subclass.
+        # More keys are prioritized over less keys (closer match). This is used when importing json to our base models not a lot of customer workflows
+        # depend on this method but this works for all our existing tests with the bonus of added validation. (no subclass found it throws an error)
+        # Previous strategies hacked but dont work for pydantic V2
+        for index in range(len(kwargs["annotations"])):
+            annotation = kwargs["annotations"][index]
+            if isinstance(annotation, dict):
+                item_annotation_keys = annotation.keys()
+                key_subclass_combos = defaultdict(list)
+                for name, subclass in subclass_registry.items():
+                    subclass: BaseModel = subclass
+                    
+                    # Get all required keys from subclass
+                    annotation_keys = []
+                    for k, field in subclass.model_fields.items():
+                        # must account for alias
+                        if hasattr(field, "validation_alias") and field.validation_alias == "answers" and "answers" in item_annotation_keys:
+                            annotation_keys.append("answers")
+                        elif field.default == PydanticUndefined and k != "uuid":
+                            annotation_keys.append(to_camel(k))
+                    key_subclass_combos[subclass].extend(annotation_keys)
 
-    class _Relationship(pydantic_compat.BaseModel):
+                # Sort by subclass that has the most keys
+                key_subclass_combos = dict(sorted(key_subclass_combos.items(), key = lambda x : len(x[1]), reverse=True))
+                
+                # Choose the key that our dict we are passing in has all the keys to match
+                for subclass, key_subclass_combo in key_subclass_combos.items():
+                    check_required_keys = all(key in list(item_annotation_keys) for key in key_subclass_combo)
+                    if check_required_keys:
+                        # Keep trying subclasses until we find one that has valid values
+                        with suppress(ValidationError):
```

Comment:
why suppressing validations? what if it does not pass validations?

Reply:
is there a good unit test to test this logic?

Reply:
So it is cycling best-fitting base models until one sticks i.e. has valid keys does not error out. I can modify it to peak at the model first and then, if valid, pick that one. But in this case, I went with the "feet-first" option it simplifies the code slightly and it a little bit faster since I would have to do another loop with to look at the keys. As soon as it finds a model that fits the JSON Schema, it breaks the loop, with the chosen model. I think this also reduces some maintenance since it has to have a model that fits the JSON schema and correctly serializes without errors. If it does not, the dictionary (JSON) stays a dictionary and errors out. 

Now that I think about it though I can do two things to help:
1. Provide better error messaging when a valid model is not found (clear messaging)
2. Write a couple of unit tests one that should be valid and one that should not be valid

What are your thoughts?

Reply:
Did the two point above!